File: PDA.FRONT/src/shared/services/master.service.ts
import {  LOT_INFORMATION_DTO, RACK_LOCATION_DTO, STORAGE_DTO, PRINTER_DTO } from "@appShared/interfaces";
import { Injectable } from "@angular/core";
import { appSettings } from "@appSettings";
import {  HTTP } from "hwmx-angular/tools"; 

@Injectable({ providedIn: 'root' })
export class MasterService extends HTTP {
 
    private readonly controller = `${appSettings.webAPI.hwmxPDA}/api/Master`;


    /** HTTP GET */
    public GetLotInformation = async (lotNumber: string) => {
        const response = await HTTP.GET<LOT_INFORMATION_DTO>({
            url: `${this.controller}/GetLotInformation/${lotNumber}`
        });

        if(response.ok) {
            return response.data;
        }

        else {
            if(response.status < 500) {
                this.alert.CloseAllAlerts();
                this.alert.Warning(response.message, lotNumber, 'barcode');
            }

            else {
                console.error(response.message);
                this.alert.Danger('GetLotInformation', 'Error', 'bug');
            }

            return null;
        }
    }


    /** HTTP GET */
    public GetCaseLabel = async (caseLabel: string, storageCode: string = '') => {
        const response = await HTTP.GET<LOT_INFORMATION_DTO[]>({
            url: `${this.controller}/GetCaseLabel/${caseLabel}`,
            queryParams: [
                { param: 'storageCode', value: storageCode }
            ]
        });

        if(!response.ok) {
            console.error(response.message);
            this.alert.Danger('GetCaseLabel', 'Error', 'bug');
            return [];
        }

        return response.data;
    }


    /** HTTP GET */
    public GetLotListByLocation = async (location: string) => {
        const response = await HTTP.GET<LOT_INFORMATION_DTO[]>({
            url: `${this.controller}/GetLotListByLocation/${location}`
        });

        if(!response.ok) {
            console.error(response.message);
            this.alert.Danger('GetLotListByLocation', 'Error', 'bug');
            return [];
        }

        return response.data;
    }


    /** HTTP GET */
    public GetStorageList = async (factory: string = '', storageType: string = '') => {
        const response = await HTTP.GET<STORAGE_DTO[]>({
            url: `${this.controller}/GetStorageList`,
            queryParams: [
                { param: 'factory',     value: factory     },
                { param: 'storageType', value: storageType }
            ]
        });

        if(!response.ok) {
            console.error(response.message);
            this.alert.Danger('GetStorageList', 'Error', 'bug');
        }

        return response.data;
    }


    /** HTTP GET */
    public GetLocation = async (location: string, showWarning: boolean = true) => {
        const response = await HTTP.GET<RACK_LOCATION_DTO>({
            url: `${this.controller}/GetLocation/${location}`
        });

        if(response.ok) {
            return response.data;
        }

        else {
            if(response.status < 500) {
                if(showWarning) {
                    this.alert.CloseAllAlerts();
                    this.alert.Warning(response.message, location, 'location');
                }
            }

            else {
                console.error(response.message);
                this.alert.Danger('GetLocation', 'Error', 'bug');
            }

            return null;
        }
    } 


    /** HTTP GET */
    public GetLocationList = async (rack: string | null, rackType: string | null) => {
        const response = await HTTP.GET<RACK_LOCATION_DTO[]>({
            url: `${this.controller}/GetLocationList`,
            queryParams: [
                { param: 'rack'    , value: rack     },
                { param: 'rackType', value: rackType }
            ]
        });

        if(!response.ok) {             
            console.error(response.message);
            this.alert.Danger('GetLocationList', 'Error', 'bug');
            
            return [];
        } 

        return response.data;
    }


    /** HTTP GET */
    public GetLocationByMaterial = async (partNumber: string) => {
        const response = await HTTP.GET<any[]>({
            url: `${this.controller}/GetLocationByMaterial/${partNumber}`
        });

        if(!response.ok) {
            if(response.status < 500) {                 
                this.alert.CloseAllAlerts();
                this.alert.Warning(response.message, partNumber, 'material');                 
            }

            else {
                console.error(response.message);
                this.alert.Danger('GetLocationByMaterial', 'Error', 'bug');
            }

            return [];
        } 

        return response.data;
    }


    /** HTTP GET */
    public GetPrinterList = async () => {
        const response = await HTTP.GET<PRINTER_DTO[]>({
            url: `${this.controller}/GetPrinterList`
        });

        if(!response.ok) {             
            console.error(response.message);
            this.alert.Danger('GetPrinterList', 'Error', 'bug');
            
            return [];
        } 

        return response.data;
    }
}

File: PDA.BACK/Microservices/Services/MasterService.cs
using Microservices.Interfaces;
using Repositories.Interfaces;
using Repositories.Database; 
using HWMX.DotNet.ORM;
using HWMX.DotNet;

namespace Microservices.Services
{
    public class MasterService(IMasterRepository _repository) : IMasterService
    {

        public async Task<ResponseDTO<LOT_INFORMATION_DTO>> GetLotInformation(string lotNumber)
        {
            ResponseDTO<LOT_INFORMATION_DTO> response = new();

            try
            {
                ResponseProcedure responseProcedure = await _repository.GetLotInformation(lotNumber);

                if (responseProcedure.Failure)
                    return response.Error(responseProcedure.MessageList);

                response.Data = responseProcedure.GetTable<LOT_INFORMATION_DTO>().FirstOrDefault();

                if (response.Data is null)
                    return response.NotFound(); 
             }

            catch (Exception ex)
            {
                return response.Exception(ex);
            }

            return response;
        } 


        public async Task<ResponseList<LOT_INFORMATION_DTO>> GetLotListByCaseLabel(string caseLabel, string storageCode = "")
        {
            ResponseList<LOT_INFORMATION_DTO> response = new();

            try
            {
                ResponseProcedure responseProcedure = await _repository.GetLotListByCaseLabel(caseLabel, storageCode);

                if (responseProcedure.Failure)
                    return response.Error(responseProcedure.MessageList);

                response.Data = responseProcedure.GetTable<LOT_INFORMATION_DTO>(); 
            }

            catch (Exception ex)
            {
                return response.Exception(ex);
            }

            return response;
        }


        public async Task<ResponseList<LOT_INFORMATION_DTO>> GetLotListByLocation(string location)
        {
            ResponseList<LOT_INFORMATION_DTO> response = new();

            try
            {
                ResponseProcedure responseProcedure = await _repository.GetLotListByLocation(location);

                if (responseProcedure.Failure)
                    return response.Error(responseProcedure.MessageList);

                response.Data = responseProcedure.GetTable<LOT_INFORMATION_DTO>(); 
            }

            catch (Exception ex)
            {
                return response.Exception(ex);
            }

            return response;
        }


        public async Task<ResponseList<STORAGE_DTO>> GetStorageList(string factory = "", string storageType = "")
        {
            ResponseList<STORAGE_DTO> response = new();

            try
            {
                ResponseProcedure responseProcedure = await _repository.GetStorageList(factory, storageType);

                if (responseProcedure.Failure)
                    return response.Error(responseProcedure.MessageList);

                response.Data = responseProcedure.GetTable<STORAGE_DTO>(); 
            }

            catch (Exception ex)
            {
                return response.Exception(ex);
            }

            return response;
        }


        public async Task<ResponseDTO<RACK_LOCATION_DTO>> GetLocation(string location)
        {
            ResponseDTO<RACK_LOCATION_DTO> response = new();

            try
            {                
                ResponseProcedure responseProcedure = await _repository.GetLocation(location, null, null);

                if (responseProcedure.Failure)
                    return response.Error(responseProcedure.MessageList);

                response.Data = responseProcedure.GetTable<RACK_LOCATION_DTO>().FirstOrDefault();

                if (response.Data is null)
                    return response.NotFound(); 
            }

            catch (Exception ex)
            {
                return response.Exception(ex);
            }

            return response;
        }


        public async Task<ResponseList<RACK_LOCATION_DTO>> GetLocationList(string rack, string rackType)
        {
            ResponseList<RACK_LOCATION_DTO> response = new();

            try
            {                
                ResponseProcedure responseProcedure = await _repository.GetLocation(null, rack, rackType);

                if (responseProcedure.Failure)
                    return response.Error(responseProcedure.MessageList);

                response.Data = responseProcedure.GetTable<RACK_LOCATION_DTO>();  
            }

            catch (Exception ex)
            {
                return response.Exception(ex);
            }

            return response;
        }


        public async Task<ResponseList<RACK_LOCATION_DTO>> GetLocationByMaterial(string partNumber)
        {
            ResponseList<RACK_LOCATION_DTO> response = new();

            try
            {
                ResponseProcedure responseProcedure = await _repository.GetLocationByMaterial(partNumber);

                if (responseProcedure.Failure)
                    return response.Error(responseProcedure.MessageList);

                response.Data = responseProcedure.GetTable<RACK_LOCATION_DTO>(); 
            }

            catch (Exception ex)
            {
                return response.Exception(ex);
            }

            return response;
        }


        public async Task<ResponseList<PRINTER_DTO>> GetPrinterList()
        {
            ResponseList<PRINTER_DTO> response = new();

            try
            {
                ResponseProcedure responseProcedure = await _repository.GetPrinterList(null);

                if (responseProcedure.Failure)
                    return response.Error(responseProcedure.MessageList);

                response.Data = responseProcedure.GetTable<PRINTER_DTO>(); 
            }

            catch (Exception ex)
            {
                return response.Exception(ex);
            }

            return response;
        } 
    }
} 